### PR TITLE
Tracking RMC4_UE5 changes vs master

### PR DESCRIPTION
**Pull request by connorjak**
This might not be merged for a long time, or ever, but it's useful to have a PR open for the reviewing tools.

Per @Moddingear: there are macro definition issues which block having a single codebase that works for both UE4 and UE5.

We should still do our best to retain compatibility in terms of making it possible to port non-UE5-specific changes back to master branch.

### CODE REVIEW

File: Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp [hunk 1]
```diff
@@ -95,7 +95,7 @@ FPrimitiveViewRelevance FRuntimeMeshComponentSceneProxy::GetViewRelevance(const
 	#if ENGINE_MINOR_VERSION >= 26
 		bool bForceDynamicPath = IsRichView(*View->Family) || IsSelected() || View->Family->EngineShowFlags.Wireframe;
 	#else
-		bool bForceDynamicPath = !IsStaticPathAvailable() || IsRichView(*View->Family) || IsSelected() || View->Family->EngineShowFlags.Wireframe;
+		bool bForceDynamicPath = IsRichView(*View->Family) || IsSelected() || View->Family->EngineShowFlags.Wireframe;
```
Comment (connorjak) [inline Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp:98]: Revert this change; this code should not get touched in UE5.

File: Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp [hunk 2]
```diff
@@ -124,7 +124,7 @@ void FRuntimeMeshComponentSceneProxy::CreateMeshBatch(FMeshBatch& MeshBatch, con
 	const bool bRenderWireframe = WireframeMaterial != nullptr;
 
 	// Decide if we should be using adjacency information for this material
-	const bool bWantsAdjacencyInfo = !bForRayTracing && !bRenderWireframe && RequiresAdjacencyInformation(Material, Section.Buffers->VertexFactory.GetType(), GetScene().GetFeatureLevel());
+	const bool bWantsAdjacencyInfo = false;//!bForRayTracing && !bRenderWireframe && RequiresAdjacencyInformation(Material, Section.Buffers->VertexFactory.GetType(), GetScene().GetFeatureLevel());
```
Comment (connorjak) [inline Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp:127]: Raytracing disabled here; needs to be reimplemented.

File: Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp [hunk 3]
```diff
@@ -242,7 +242,7 @@ void FRuntimeMeshComponentSceneProxy::GetDynamicMeshElements(const TArray<const
 			#if ENGINE_MINOR_VERSION >= 26
```
Comment (connorjak) [inline Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp:242]: Again needs check for MAJOR_VERSION

File: Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp [hunk 4]
```diff
@@ -318,50 +318,50 @@ void FRuntimeMeshComponentSceneProxy::GetDynamicMeshElements(const TArray<const
...
-			}
-		}
-	}
+	//				RayTracingInstance.Materials.Add(MeshBatch);
+	//				RayTracingInstance.BuildInstanceMaskAndFlags();
+	//				OutRayTracingInstances.Add(RayTracingInstance);
+	//			}
+	//		}
+	//	}
+	//}
```
Comment (connorjak) [inline Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp:364]: This whole raytracing function is commented out; need to revert and make appropriate upgrades.

File: Source/RuntimeMeshComponent/Private/RuntimeMeshRendering.h [hunk 5]
```diff
@@ -200,34 +200,34 @@ class FRuntimeMeshVertexBuffer : public FVertexBuffer
...
 		CreateInfo.bWithoutNativeResource = SizeInBytes <= 0;
 		if (bIsInRenderThread)
 		{
-			return RHICreateVertexBuffer(SizeInBytes, Flags, CreateInfo);
+			return RHICreateVertexBuffer(SizeInBytes, (bDynamicBuffer ? BUF_Dynamic : BUF_Static) | BUF_ShaderResource, CreateInfo);
 		}
 		else
 		{
-			return RHIAsyncCreateVertexBuffer(SizeInBytes, Flags, CreateInfo);
+			return RHIAsyncCreateVertexBuffer(SizeInBytes, (bDynamicBuffer ? BUF_Dynamic : BUF_Static) | BUF_ShaderResource, CreateInfo);
```
Comment (connorjak) [inline Source/RuntimeMeshComponent/Private/RuntimeMeshRendering.h:213]: TODO investigate relation to #230

File: Source/RuntimeMeshComponent/Private/RuntimeMeshRendering.h [hunk 6]
```diff
@@ -548,35 +548,35 @@ class FRuntimeMeshIndexBuffer : public FIndexBuffer
...
 		CreateInfo.bWithoutNativeResource = SizeInBytes <= 0;
 		if (bIsInRenderThread)
 		{
-			return RHICreateIndexBuffer(IndexSize, SizeInBytes, Flags, CreateInfo);
+			return RHICreateIndexBuffer(IndexSize, SizeInBytes, (bDynamicBuffer ? BUF_Dynamic : BUF_Static) | BUF_ShaderResource, CreateInfo);
 		}
 		else
 		{
-			return RHIAsyncCreateIndexBuffer(IndexSize, SizeInBytes, Flags, CreateInfo);
+			return RHIAsyncCreateIndexBuffer(IndexSize, SizeInBytes, (bDynamicBuffer ? BUF_Dynamic : BUF_Static) | BUF_ShaderResource, CreateInfo);
```
Comment (connorjak) [inline Source/RuntimeMeshComponent/Private/RuntimeMeshRendering.h:561]: TODO investigate relation to #230

File: Source/RuntimeMeshComponent/Private/RuntimeMeshSectionProxy.cpp [hunk 7]
```diff
@@ -73,6 +73,6 @@ void FRuntimeMeshSectionProxyBuffers::UpdateRayTracingGeometry()
 #endif
 
 		RayTracingGeometry.UpdateRHI();
-	}
+	}*/
```
Comment (connorjak) [inline Source/RuntimeMeshComponent/Private/RuntimeMeshSectionProxy.cpp:76]: Again revert comment-out of raytracing code and properly update.

File: Source/RuntimeMeshComponent/Public/RuntimeMeshCollision.h [hunk 8]
```diff
@@ -734,7 +734,7 @@ struct RUNTIMEMESHCOMPONENT_API FRuntimeMeshCollisionData
 
 	bool HasValidMeshData()
 	{
-		return Vertices.Num() >= 3 && Triangles.Num() >= 3;
+		return Vertices.Num() >= 3 && Triangles.Num() >= 1;
```
Comment (connorjak) [inline Source/RuntimeMeshComponent/Public/RuntimeMeshCollision.h:737]: TODO validate change

**Review by connorjak — COMMENTED** [review]
Some slight issues but the major one is Raytracing functionality was commented out. Will need to fix.

File: Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp [hunk 9]
```diff
@@ -242,7 +242,7 @@ void FRuntimeMeshComponentSceneProxy::GetDynamicMeshElements(const TArray<const
 			#if ENGINE_MINOR_VERSION >= 26
 				bool bForceDynamicPath = IsRichView(*Views[ViewIndex]->Family) || Views[ViewIndex]->Family->EngineShowFlags.Wireframe || IsSelected();
 			#else
-				bool bForceDynamicPath = IsRichView(*Views[ViewIndex]->Family) || Views[ViewIndex]->Family->EngineShowFlags.Wireframe || IsSelected() || !IsStaticPathAvailable();
+				bool bForceDynamicPath = IsRichView(*Views[ViewIndex]->Family) || Views[ViewIndex]->Family->EngineShowFlags.Wireframe || IsSelected();
```
[the remark this answers is not in the document]
Comment (connorjak) [reply Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp:245]: #232

File: Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp (hunk 3 above)
Comment (connorjak) [reply Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp:242]: #232

File: Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp (hunk 1 above)
Comment (connorjak) [reply Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp:98]: #232

File: Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp [hunk 10]
```diff
@@ -95,7 +95,7 @@ FPrimitiveViewRelevance FRuntimeMeshComponentSceneProxy::GetViewRelevance(const
 	#if ENGINE_MINOR_VERSION >= 26
```
[the remark this answers is not in the document]
Comment (connorjak) [reply Source/RuntimeMeshComponent/Private/RuntimeMeshComponentProxy.cpp:95]: #232

File: Source/RuntimeMeshComponent/Private/RuntimeMeshActor.cpp [hunk 11]
```diff
@@ -10,10 +10,11 @@
...
 {
-#if ENGINE_MAJOR_VERSION >= 4 && ENGINE_MINOR_VERSION >= 24
-	SetCanBeDamaged(false);
-#else
+
+#if ENGINE_MAJOR_VERSION <= 4 && ENGINE_MINOR_VERSION <= 24
 	bCanBeDamaged = false;
+#else
+	SetCanBeDamaged(false);
 #endif
```
[the remark this answers is not in the document]
Comment (connorjak) [reply Source/RuntimeMeshComponent/Private/RuntimeMeshActor.cpp:18]: #232

**Comment by Koderz** [comment]
Closing as this won't apply to the new version. v5 is dropping backwards compatibility with UE4 due largely to this, but also to some other renderer changes and future plans.
